### ISSUE TITLE
[Feature] Allow test.step to return the value resolved in the test callback

### ISSUE THREAD

**Issue by JCMais** [microsoft/playwright#9742]
Right now `test.step` resolves to `undefined`.

It would be great if it resolved to the value the callback resolves to, so we could have tests like this:

```ts
test('test something', async ({ pageObject, page }) => {
  await pageObject.navigateTo();
  
  const link = await test.step('is able to retrieve a link', async () => {
    const link = await pageObject.getLink();
    
    // tests...
    expect(link).toBe(blabla);
    
    return link;
  });
  
  await test.step('link is accessible', async () => {
    await page.goto(link)
    // tests...
    expect(page.locator('#my-element')).toBeVisible();
  });
})
```

Right now the workaround is to have `let` variables before the step (notice the `!` to avoid type errors):

```ts
test('test something', async ({ pageObject, page }) => {
  await pageObject.navigateTo();
  
  let link!: string
  await test.step('is able to retrieve a link', async () => {
    link = await pageObject.getLink();
    
    // tests...
    expect(link).toBe(blabla);
  });
  
  await test.step('link is accessible', async () => {
    await page.goto(link)
    // tests...
    expect(page.locator('#my-element')).toBeVisible();
  });
})
```

I'm free to work on this if it is something that would be accepted in the library.

**Comment by dgozman** [comment]
@JCMais Thank you for the PR! We discussed this proposal, and are not ready to commit to it just yet - there are various other ideas of what `test.step()` could  return. I've changed this to P3, let's see whether this would be a common request.

**Comment by hananmalka** [comment]
@dgozman - I support this as a must have feature. 
I'm wrapping all my steps with test.step to be able to customize allure report steps. 
When doing that - I'm not able to return any value from my methods. 
test.step should be able to support a return value.

**Comment by jetwiwo** [comment]
I'd really like to see this ability. I quite often will want to write a number of steps where one step uses the result of the previous for my tests. Creating a local variable works, but then you have to explicitly type that variable every time which is tedious. Having the test.step() return the value would allow me to use type inference (e.g. `const step1Result = await test.step(...)`). If test.step() is to return "something else", that something else could include the result instead.

Perhaps while the decision is being made, a reasonable interim compromise would be to have test.step be something like

```typescript
step<RESULT_TYPE = any>(description: string, callback: () => Promise<RESULT_TYPE>): Promise<{result: RESULT_TYPE}>
```

This way if you want to return other things as well, you can just add extra properties.

**Comment by ozyx** [comment]
Possible duplicate of https://github.com/microsoft/playwright/issues/15614 ?

**Comment by dgozman** [comment]
Yep, and implemented in #16060.
